### PR TITLE
Allow passing arguments containing spaces into pygmt functions

### LINKED ISSUE #247: Arguments with spaces in strings
**Description of the problem**

**Full code that generated the error**

I tried to plot a basemap with an axis labelled as "Crust Age". I tried in three different ways:
```python
import gmt

fig = gmt.Figure()
# 1st way: FAIL
fig.basemap(R=[0, 100, 0, 50], J="X10c/5c", B='x20+lCrust Age')
# 2nd way: SUCCEED
fig.basemap(R=[0, 100, 0, 50], J="X10c/5c", B='x20+l"Crust Age"')
# 3rd way: FAIL
fig.basemap(R=[0, 100, 0, 50], J="X10c/5c", B="x20+l'Crust Age'")
fig.savefig("map.pdf")
```

**Full error message**

```
psbasemap [ERROR]: Unrecognized option -<
Traceback (most recent call last):
  File "test.sh", line 6, in <module>
    fig.basemap(R=[0, 100, 0, 50], J="X10c/5c", B="x20+l'Crust Age'")
  File "/Users/seisman/Gits/gmt/gmt-python/gmt/helpers/decorators.py", line 199, in new_module
    return module_func(*args, **kwargs)
  File "/Users/seisman/Gits/gmt/gmt-python/gmt/helpers/decorators.py", line 294, in new_module
    return module_func(*args, **kwargs)
  File "/Users/seisman/Gits/gmt/gmt-python/gmt/base_plotting.py", line 472, in basemap
    lib.call_module("basemap", build_arg_string(kwargs))
  File "/Users/seisman/Gits/gmt/gmt-python/gmt/clib/session.py", line 490, in call_module
    module, status, self._error_message
gmt.exceptions.GMTCLibError: Module 'basemap' failed with status code 71:
psbasemap [ERROR]: Unrecognized option -<
```

The 1st way is equivalent to command line

    gmt basemap -R0/100/0/50 -JX10c/5c -Bx20+lCrust Age 

The command fails because "Age" is thought to be a file. However, the error message is almost useless to help users find the error. Do you think it's technically possible to automatically add quotes around strings with spaces if users forget to do that?

The 2nd way works as expected. However, I don't understand why the 3rd way fails? The equivalent cmd version works:
```
gmt basemap -R0/100/0/50 -JX10c/5c -Bx20+l'Crust Age' -pdf map
```

**System information**

* Operating system: macOS
* Python installation (Anaconda, system, ETS): Anaconda
* Version of GMT: 6.0.0_15a9160
* Version of Python: 3.7.0
* Version of this package: latest

### PR DESCRIPTION
**Pull request by weiji14**
**Description of proposed changes**

To allow for spaces in arguments to PyGMT functions, this pull request modifies the `build_arg_string` function in `pygmt/helpers/utils.py` to replace space characters (` `) with the equivalent octal code (`\040`). This enables users to do this:

```python
import pygmt

fig = pygmt.Figure()
fig.basemap(region=[0, 2, 4, 8], projection="X2c", frame=r"WSne+tPlot Zero")
fig.show()
```

Note that no nested double quotes (`'"some word"'`) are needed anymore!

I have also removed some hacky workarounds that allowed for text inputs with spaces into:
- `subplot`'s `autolabel`(-A) and `title` (-T) parameters - workaround added in #1125 and #822
- `text`'s `-F` parameter
- `psconvert`'s `prefix` (-F) parameter - workaround added in #1116

Also updated some unit tests for `fig.legend` and `fig.rose` that had used arguments like `'"Some text"'`.

### **Special handling** of functions where spaces can't just be converted to `\040`:
- [x] Ensure that `pygmt.config(FORMAT_DATE_MAP="o dd")` works (c29e63229f6c46aeb47cfe068b9f6c959728f64a) 
- [x] Make sure that `fig.savefig()` saves filenames with spaces properly (#1116) without `\040` octal code characters in the filename (83c8c3b0e8c699cd52e7d539c49dabbbf7bf384c)
- [x] Ensure that PROJ4 projections with spaces (e.g. `projection="+proj4=+longlat +ellipsoid=someplanet"`) work, xref https://github.com/GenericMappingTools/pygmt/issues/655#issuecomment-713154812 (3ec772740823a12f906210edeed5284aa101ff12)

Fixes #247 :tada: :tada: :tada:

**Reminders**

- [x] Run `make format` and `make check` to make sure the code follows the style guide.
- [x] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [x] Write detailed docstrings for all functions/methods.
- [ ] If adding new functionality, add an example to docstrings or tutorials.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

File: pygmt/tests/test_figure.py [hunk 1]
```diff
@@ -116,6 +116,7 @@ def test_figure_savefig_filename_with_spaces():
     fig.basemap(region=[0, 1, 0, 1], projection="X1c/1c", frame=True)
     with GMTTempFile(prefix="pygmt-filename with spaces", suffix=".png") as imgfile:
         fig.savefig(imgfile.name)
+        assert r"\040" not in os.path.abspath(imgfile.name)
         assert os.path.exists(imgfile.name)
```
Comment (weiji14) [inline pygmt/tests/test_figure.py:120]: A problem with this PR is that running `fig.savefig` with filenames that have spaces (introduced in #1116) will produce an output like "pygmt-filename\040with\040spaces.png". This test currently passes still, but it is a **false positive**, because Python's `os.path.exists` reads "pygmt-filename\040with\040spaces.png" as ""pygmt-filename with spaces.png"".

Any ideas how to improve the test and/or implementation of `psconvert`?

**Comment by seisman** [comment]
xref: https://forum.generic-mapping-tools.org/t/errors-in-pygmt-when-there-are-spaces-in-string-arguments/2494/3

File: pygmt/src/config.py [hunk 2]
```diff
@@ -55,7 +55,7 @@ def __init__(self, **kwargs):
                     self.old_defaults[key] = lib.get_default(key)
 
         # call gmt set to change GMT defaults
-        arg_str = " ".join([f"{key}={value}" for key, value in kwargs.items()])
+        arg_str = " ".join([f'{key}="{value}"' for key, value in kwargs.items()])
```
Comment (weiji14) [inline pygmt/src/config.py:58]: > xref: https://forum.generic-mapping-tools.org/t/errors-in-pygmt-when-there-are-spaces-in-string-arguments/2494/3

Ok, I've fixed the `FORMAT_DATE_MAP="o dd"` issue reported in the forum by wrapping the arguments in double quotes here in commit c29e63229f6c46aeb47cfe068b9f6c959728f64a. Didn't use `\040` because `o\040dd` doesn't work. The problem with this solution is that the workaround suggested in the forum (use `FORMAT_DATE_MAP='"o dd"'`) will break in PyGMT v0.6.0, but I think that's acceptable since that workaround isn't intuitive anyway.

File: pygmt/tests/test_rose.py [hunk 3]
```diff
@@ -152,7 +152,7 @@ def test_rose_no_sectors(data_fractures_compilation):
         region=[0, 500, 0, 360],
         diameter="10c",
         labels="180/0/90/270",
-        frame=["xg100", "yg45", "+t'Windrose diagram'"],
+        frame=["xg100", "yg45", "+tWindrose diagram"],
```
Comment (seisman) [inline pygmt/tests/test_rose.py:155]: Does `"+t'Windrose diagram'"` still work or not?

Comment (weiji14) [reply pygmt/tests/test_rose.py:155]: Yes it does still work, but the single quotes would be printed. I think it was a typo from whoever wrote that test.

**Comment by willschlitzer** [comment]
As far as I can tell, this added function seems to be working. Am I missing some remaining issues before it will be approved?

**Comment by weiji14** [comment]
> As far as I can tell, this added function seems to be working. Am I missing some remaining issues before it will be approved?

Yes it works. But I have a nagging feeling that there might be some edge cases that I haven't handled. Actually, this one 👇🏻 still needs a unit test

＞Ensure that PROJ4 projections with spaces (e.g. projection="+proj4=+longlat +ellipsoid=someplanet") work, xref -JEPSG:4326 not writing CRS to netcdf in surface #655 (comment)

**Comment by weiji14** [comment]
> > As far as I can tell, this added function seems to be working. Am I missing some remaining issues before it will be approved?
> 
> Yes it works. But I have a nagging feeling that there might be some edge cases that I haven't handled. Actually, this one 👇🏻 still needs a unit test
> 
>     * [ ]  Ensure that PROJ4 projections with spaces (e.g. projection="+proj4=+longlat +ellipsoid=someplanet") work, xref -JEPSG:4326 not writing CRS to netcdf in surface [-JEPSG:4326 not writing CRS to netcdf in surface #655](https://github.com/GenericMappingTools/pygmt/issues/655) (comment)

Ok, PROJ4 strings with spaces now work as of 3ec772740823a12f906210edeed5284aa101ff12 :tada: I just converted deleted the spaces for -J (projection) and it works, since GMT allows it (xref https://github.com/GenericMappingTools/gmt/blob/c697c355f39619911ffca266788e818c6ffce73e/doc/rst/source/explain_-Jproj_full.rst_#L9-L11)

File: pygmt/figure.py [hunk 4]
```diff
@@ -235,10 +235,14 @@ def psconvert(self, icc_gray=False, **kwargs):
                 kwargs["N"] = "+i"
             else:
                 kwargs["N"] += "+i"
-        # allow for spaces in figure name
-        kwargs["F"] = f'"{kwargs.get("F")}"' if kwargs.get("F") else None
+
+        # Manually handle prefix -F argument so spaces aren't converted to \040
+        # by build_arg_string function. For more information, see
+        # https://github.com/GenericMappingTools/pygmt/pull/1487
+        prefix = kwargs.pop("F")
```
Comment (meghanrjones) [inline pygmt/figure.py:242]: Should there be a check before this that ``prefix`` is provided so that the user doesn't get a KeyError here?

**Comment by meghanrjones** [comment]
I tried for a while to break it. Nice work on this complicated issue! 

The two issues that I came across relate to the ``outgrid`` and ``outfile`` parameters.

#### Outgrid parameter

The following will produce a filename with spaces using the main branch but the filename contains "\040" with this branch:
```
pygmt.grdcut("@earth_relief_01d", region=[0, 1, 0, 1], outgrid="test grid output.nc")
```

#### Outfile parameter

File outputs with spaces still require using the trick of wrapping double quotes in single quotes (this could be addressed separately from this PR).

```
pygmt.blockmean(data="@ship_15.txt", region=[245, 255, 20, 30], spacing="5m", outfile='"Test file output.txt"') # Works
pygmt.blockmean(data="@ship_15.txt", region=[245, 255, 20, 30], spacing="5m", outfile="Test file output.txt") # Doesn't work
```

File: pygmt/figure.py (hunk 4 above)
Comment (weiji14) [reply pygmt/figure.py:242]: Good point. Looking at the docs at https://docs.generic-mapping-tools.org/6.3/psconvert.html, it reads like `prefix`/`-F` is optional, but when I tried running `psconvert` without the `-F` option, it throws an error: `psconvert [ERROR]: Modern GMT mode requires the -F option`. See https://github.com/GenericMappingTools/gmt/blob/adb244afa51ca7246cc051080c9d47193087d6c2/src/psconvert.c#L1041-L1042

So since PyGMT is modern mode only, it should be fine to keep this line intact.

**Comment by weiji14** [comment]
> I tried for a while to break it. Nice work on this complicated issue!

Thanks for testing this! I still have a nagging feeling that there's a random edge case that will break this, but we'll find out once this gets released into the wild :smiling_imp:

> The two issues that I came across relate to the `outgrid` and `outfile` parameters.
> #### Outgrid parameter
> 
> The following will produce a filename with spaces using the main branch but the filename contains "\040" with this branch:
> 
> ```
> pygmt.grdcut("@earth_relief_01d", region=[0, 1, 0, 1], outgrid="test grid output.nc")
> ```
> 
> #### Outfile parameter
> 
> File outputs with spaces still require using the trick of wrapping double quotes in single quotes (this could be addressed separately from this PR).
> 
> ```
> pygmt.blockmean(data="@ship_15.txt", region=[245, 255, 20, 30], spacing="5m", outfile='"Test file output.txt"') # Works
> pygmt.blockmean(data="@ship_15.txt", region=[245, 255, 20, 30], spacing="5m", outfile="Test file output.txt") # Doesn't work
> ```

Right, fixing this would require a workaround like the one applied to `fig.savefig` mentioned in https://github.com/GenericMappingTools/pygmt/pull/1487#discussion_r783577450, but it'll be a lot of work to apply it individually to all the functions having an outgrid/outfile parameter. Maybe leave it for another time, and honestly, we really need to discourage people from having spaces in filenames in the first place.

File: pygmt/figure.py (hunk 4 above)
Comment (meghanrjones) [reply pygmt/figure.py:242]: The docs say "If no input files are given, will convert the current active figure (see pygmt.figure). In this case, an output name must be given using parameter prefix." How does one provide other input files than the current figure using pygmt.Figure.psconvert?

If we chose not to support operations analogous to ``psconvert -Tg test.ps`` (which uses the same prefix as the original file name), I think we should make ``prefix`` required in the function signature or raise an invalid input exception to avoid the traceback below. I don't think it's obvious to that the error is caused by not using ``prefix``.

image

**Review by meghanrjones — APPROVED** [review]
There are still a couple issues to work out as mentioned in https://github.com/GenericMappingTools/pygmt/pull/1487/files#r825371307 and https://github.com/GenericMappingTools/pygmt/pull/1487#issuecomment-1065965868, but this is a big improvement that would benefit users in v0.6.0.

File: pygmt/figure.py (hunk 4 above)
Comment (weiji14) [reply pygmt/figure.py]: > How does one provide other input files than the current figure using pygmt.Figure.psconvert?

We can't, because it has not been implemented yet. The current `fig.psconvert` is a method of the `pygmt.Figure()` class and is tied to the `fig`, and can only 'psconvert' the active figure.

> If we chose not to support operations analogous to `psconvert -Tg test.ps` (which uses the same prefix as the original file name), I think we should make `prefix` required in the function signature or raise an invalid input exception to avoid the traceback below. I don't think it's obvious to that the error is caused by not using `prefix`.

Ok, I've added a try-except block in 801ba01b3c83a134f6dd15f756223f8e3663513a to raise a GMTInvalidInput if `prefix`/`-F` is not given.

**Comment by seisman** [comment]
The one thing that makes little sense to me is shown in the following example:
```
import pygmt
fig = pygmt.Figure()
fig.basemap(region=[0, 2, 4, 8], projection="X10c", frame=["WSne+tPlot Zero", "xaf+lX label+uK"])
fig.show()
```
As you can see, I have to use `"xaf+lX label+uK"`, not `"xaf+l'X label'+uK"` or `'xaf+l"X label"+uK'`. For me, the last two are much easier to read.

**Comment by weiji14** [comment]
> As you can see, I have to use `"xaf+lX label+uK"`, not `"xaf+l'X label'+uK"` or `'xaf+l"X label"+uK'`. For me, the last two are much easier to read.

I just tested all three examples and only `"xaf+l'X label'+uK"` doesn't work properly (it prints `'X label` instead of `X label`), but that's a know issue with using single quotes. This branch should be backwards compatible with the `'xaf+l"X label"+uK'` method, and it now allows for `"xaf+lX label+uK"` which isn't nice but works if people like that format.

**Comment by seisman** [comment]
> > As you can see, I have to use `"xaf+lX label+uK"`, not `"xaf+l'X label'+uK"` or `'xaf+l"X label"+uK'`. For me, the last two are much easier to read.
> 
> I just tested all three examples and only `"xaf+l'X label'+uK"` doesn't work properly (it prints `'X label` instead of `X label`), but that's a known issue with using single quotes. This branch should be backwards compatible with the `'xaf+l"X label"+uK'` method, and it now allows for `"xaf+lX label+uK"` which isn't nice but works if people like that format.

Ah, you're right. Good to know that I can still use `'xaf+l"X label"+uK'`. 

I'm reading how GMT parses a long text string  https://github.com/GenericMappingTools/gmt/blob/434ed1848a1e8886931694b2de4c352e4c6cddae/src/gmt_parse.c#L600, it seems it ignores double quotes (that's why `'xaf+l"X label"+uK'`), but doesn't ignore single quotes (so `"xaf+l'X label'+uK"` doesn't work).

File: pygmt/helpers/utils.py [hunk 5]
```diff
@@ -131,6 +131,11 @@ def build_arg_string(kwargs):
     same command line argument. For example, the kwargs entry ``'B': ['xa',
     'yaf']`` will be converted to ``-Bxa -Byaf`` in the argument string.
 
+    Note that spaces ` ` in arguments are converted to the equivalent octal
+    code `\040`, except in the case of -J (projection) arguments where PROJ4
```
Comment (seisman) [inline pygmt/helpers/utils.py:135]: It's ReST, not markdown.
```suggestion
    Note that spaces `` `` in arguments are converted to the equivalent octal
    code ``\040``, except in the case of -J (projection) arguments where PROJ4
```

File: pygmt/helpers/utils.py [hunk 6]
```diff
@@ -185,11 +200,19 @@ def build_arg_string(kwargs):
...
             gmt_args.append(f"-{key}")
         else:
-            gmt_args.append(f"-{key}{kwargs[key]}")
+            if key != "J":  # non-projection parameters
+                _value = str(kwargs[key]).replace(" ", r"\040")
+            else:
+                # special handling if key == "J" (projection)
+                # remove any spaces in PROJ4 string
+                _value = str(kwargs[key]).replace(" ", "")
+            gmt_args.append(rf"-{key}{_value}")
```
Comment (seisman) [inline pygmt/helpers/utils.py:215]: I know almost nothing about PROJ4, but does a PROJ4 string always contains `+proj=xxx` or `EPSG`? If yes, then we can simplify the if-test to something like:
```
            if key != "J" or (key == "J" and "+proj" not in value and "EPSG" not in value):
                _value = str(kwargs[key]).replace(" ", r"\040")
            else:
                # special handling if -J + PROJ4 string
                # remove any spaces in PROJ4 string
                _value = str(kwargs[key]).replace(" ", "")
            gmt_args.append(rf"-{key}{_value}")
```
The motivation is, since PROJ4 is not commonly used, it seems a waste of time to do call the `replace` method for GMT-style -J.

Comment (weiji14) [reply pygmt/helpers/utils.py:215]: Not sure if they all have either `+proj=xxx` or `EPSG` (see https://proj.org/operations/projections/index.html), but they probably do? I know that there's `+init=EPSG:xxx` which works when I tried here, but that's actually deprecated syntax (see https://pyproj4.github.io/pyproj/stable/gotchas.html#init-auth-auth-code-should-be-replaced-with-auth-auth-code) and not probably something we want to allow.

Also, not sure if 2 extra if-checks saves time, considering that either way, a `str.replace` still needs to happen for the space character - either to `\040` or to nothing ``.

**Review by seisman — APPROVED** [review]

**Comment by weiji14** [comment]
Thanks team! I'll merge this in and close that almost 4 year old #247 bug!!
